Windows common libraries and install sample (cmake)

chcosta opened the pull request:
Windows implementation of native toolset bootstrapping, https://github.com/dotnet/arcade/pull/131 (I'll update the doc to reflect the implementation tomorrow).

@weshaggard @natemcmaster @tmat, can you take a look at the direction of the native toolset bootstrapping implementation and let me know if anything stands out as problematic for your repos? 

I intend to implement similar libraries / structures for Linux.  There are still some unanswered design questions like version detection.

The entry point is `eng/common/nativetoolsbootstrap.cmd` which parses `NativeToolsVersions.txt` and then executes related installer (ie `eng/common/native/install-cmake.ps1` )

markwilkie commented:
What about using hash compare for version detection since we don't need to determine version order?

alexperovich commented on eng/common/native/CommonLibraryDownloadAndExtract.psm1:
```diff
@@ -0,0 +1,73 @@
+Helper module to install an xcopy deployable native tool
+
+.DESCRIPTION
+Helper module to install an xcopy deployable native tool
```
NIT: Synopsis and Description should not be identical

alexperovich commented on eng/common/native/CommonLibraryDownloadAndExtract.psm1:
```diff
@@ -0,0 +1,73 @@
+    $ToolFilename = Split-Path $Url -leaf
+    $TempToolPath = Join-Path $TempDir $ToolFilename
+
+    # Download native tool
```
NIT: Add some Write-Verbose calls here to enable `-Verbose` logging of what this script is doing.

alexperovich commented on eng/common/native/CommonLibraryDownloadWithRetry.psm1:
```diff
@@ -0,0 +1,76 @@
+            $Success = $True
+        }
+        catch {
+            Write-Host $_
```
Should Write-Warning for failed attempts, with the last attempt error using a Write-Error

alexperovich commented on eng/common/native/CommonLibraryGenerateShim.psm1:
```diff
@@ -0,0 +1,60 @@
+        }
+
+        $ShimContents = "@echo off`n"
+        $ShimContents += "$AssemblyPath %*"
```
I know this specific shim contents doesn't work well at all with parameters that have spaces and/or other things that need quotes in them. What should we do for that case.

chcosta replied:
Any suggestions or examples of where this has been handled previously.

Are you thinking something like this would be preferred?

```
@echo off
setlocal enableextensions enabledelayedexpansion

set args=
for %%x in (%*) do (
    set args=!args! "%%~x"
)
foo.cmd %args%
```

alexperovich replied:
I am not a cmd expert. That may work. In bash it would just be `foo.sh "$@"`.
When I have needed shims on windows I have found it easier to use a C# app I wrote that uses roslyn to produce an exe that uses Process.Start and constructs the args correctly. That's probably way more than we need here.

natemcmaster replied:
We've used `%*` with reasonable success in aspnet, but yes, it is still a pain. I haven't found anything better since looping over args doesn't preserve arguments with `=` in it, for whatever reason.

alexperovich commented on eng/common/native/CommonLibraryGenerateShim.psm1:
```diff
@@ -0,0 +1,60 @@
+        $ShimContents += "$AssemblyPath %*"
+        
+        # Write shim file
+        $ShimContents | Out-File $ShimPath -Encoding "ASCII"
```
Maybe Write-Verbose the stuff we are about to write to the file?

alexperovich commented on eng/common/native/CommonLibraryModules.psm1:
```diff
@@ -0,0 +1,8 @@
+# Import this module to import all of the common libraries into your PS script
```
Why not put all these functions in a single `CommonLibrary.psm1` file?

natemcmaster replied:
Btw, if you're concerned about public/private functions, you might also consider adding a `.psd1` file.

weshaggard commented:
>What about using hash compare for version detection since we don't need to determine version order?

I don't understand how this would help? This would only work if you want to do exact matches. Generally we are going to want to ask the question do we already have a this version or higher.

natemcmaster commented on eng/common/native/CommonLibraryDownloadWithRetry.psm1:
```diff
@@ -0,0 +1,76 @@
+        New-Item -path $DownloadDirectory -force -itemType "Directory" | Out-Null
+    }
+
+    Write-Verbose "Downloading $Url"
```
Consider checking first if `$Url` is a file path. I've found it's really useful for testing and local builds if you can swap out a folder path and a remote http:// endpoint.

natemcmaster commented on eng/common/native/CommonLibraryDownloadWithRetry.psm1:
```diff
@@ -0,0 +1,76 @@
+    while((-Not $Success) -And ($Attempt -Lt $DownloadRetries) )
+    {
+        try {
+            Invoke-WebRequest -Uri $Url -OutFile $Path
```
Add `-UseBasicParsing`. This is required for some installations of Windows Server.

natemcmaster commented on eng/common/nativetoolsbootstrap.ps1:
```diff
@@ -0,0 +1,129 @@
+[CmdletBinding(PositionalBinding=$false)]
+Param (
+    [string] $AzureStorageUrl = "https://dotnetfeed.blob.core.windows.net",
+    [string] $AzureContainerName = "chcosta-test",
```
I assume this is a placeholder ;)

chcosta replied:
Yes, though we should figure out a container name.

natemcmaster commented on eng/common/nativetoolsbootstrap.cmd:
```diff
@@ -0,0 +1,3 @@
+@echo off
+powershell -ExecutionPolicy ByPass %~dp0nativetoolsbootstrap.ps1 %*
```
Also add `-NoProfile`, and consider adding `-NoLogo`.

natemcmaster commented on eng/common/native/CommonLibraryDownloadAndExtract.psm1:
```diff
@@ -0,0 +1,73 @@
+.NOTES
+Returns False if download or extraction fail, True otherwise
+#>
+function CommonLibraryDownloadAndExtract {
```
Nit: about naming. Can we try to follow powershell conventions on function names? This one might be called `Extract-RemoteFile`. If you're worried about conflicts with existing powershell cmdlets, we can [fully qualified](https://www.sapien.com/blog/2015/10/23/using-module-qualified-cmdlet-names/) names to invoke the right cmdlet.

chcosta replied:
I initially started following powershell convention on function names, but quickly became annoyed with the limited set of "approved verbs".  If a verb isn't approved, you get a warning that shows up in Powershell editors or in verbose output.  So, I moved away from that convention so that I could just name the functions whatever made sense to me.  Do you think it's worth trying to massage these methods (especially combination methods like "DownloadAndExtract" into the set of [approved verbs](https://msdn.microsoft.com/en-us/library/ms714428(v=vs.85).aspx)?

Thanks for the link to "fully qualified names", I was intentionally trying to avoid conflicts with the method names, but I'll use fully qualified names instead.

natemcmaster replied:
Sounds like you've hit my least favorite PowerShell linter warning. The approved verbs are pretty limited. If possible, use one. Otherwise, pick one that makes sense. But IMO let's still use the `Verb-Noun` convention if we can.

natemcmaster commented on eng/common/native/CommonLibraryDownloadWithRetry.psm1:
```diff
@@ -0,0 +1,76 @@
+.PARAMETER Force
+Overwrite existing file if present
+#>
+function CommonLibraryDownloadWithRetry {
```
Naming suggestion: `Get-RemoteFileWithRetry`

natemcmaster commented on eng/common/native/CommonLibraryDownloadWithRetry.psm1:
```diff
@@ -0,0 +1,76 @@
+        try {
+            Invoke-WebRequest -Uri $Url -OutFile $Path
+            Write-Verbose "Downloaded to '$Path'"
+            $Success = $True
```
`return $true`

natemcmaster commented on eng/common/native/CommonLibraryGenerateShim.psm1:
```diff
@@ -0,0 +1,60 @@
+.NOTES
+Returns $True if generating shim succeeds, $False otherwise
+#>
+function CommonLibraryGenerateShim {
```
Naming suggestion: `New-ScriptShim`

natemcmaster commented on eng/common/native/CommonLibraryGenerateShim.psm1:
```diff
@@ -0,0 +1,60 @@
+    try {
+        Write-Verbose "Generating '$ShimPath' shim"
+
+        if ((Test-Path $ShimPath) -And ($Force -Eq $False)) {
```
Nit: `-not $Force`

natemcmaster commented on eng/common/native/CommonLibraryGenerateShim.psm1:
```diff
@@ -0,0 +1,60 @@
+        # Write shim file
+        $ShimContents | Out-File $ShimPath -Encoding "ASCII"
+
+        if ($? -Ne $True) {
```
Nit: `-not $?`

natemcmaster commented on eng/common/native/CommonLibraryGetArchitecture.psm1:
```diff
@@ -0,0 +1,24 @@
+  Returns 'x64' on 64 bit machines
+  Returns 'x86' on 32 bit machines
+#>
+function CommonLibraryGetArchitecture {
```
Naming suggestion: `Get-MachineArchitecture`

natemcmaster commented on eng/common/native/CommonLibraryUnzip.psm1:
```diff
@@ -0,0 +1,66 @@
+<#
+.SYNOPSIS
+Unzip an archive
```
What do you think about requiring powershell 5 and just using the built-in [`Expand-Archive`](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.archive/expand-archive?view=powershell-5.0) cmdlet?

natemcmaster commented on eng/common/native/CommonLibraryUnzip.psm1:
```diff
@@ -0,0 +1,66 @@
+- Returns False if Overwrite is True and it is unable to remove contents of OutputDirectory
+- Returns False if unable to extract zip archive
+#>
+function CommonLibraryUnzip {
```
Naming suggestion: `Extract-Zip`

chcosta replied:
`Extract` is not approved.  This would be `Expand-Zip`

natemcmaster commented on eng/common/nativetoolsbootstrap.ps1:
```diff
@@ -0,0 +1,129 @@
+.PARAMETER AzureStorageUrl
+Url to Azure blob storage containing native assets
+
+.PARAMETER AzureContainerName
```
What's the thinking on having this parameter separated out? Any reason you can't just make the storage url a base-path for downloading tools?

natemcmaster commented on eng/common/nativetoolsbootstrap.ps1:
```diff
@@ -0,0 +1,129 @@
+        Write-Host "No native tool dependencies are defined in '$ToolsForInstallFile'"
+        exit 0
+    }
+    $ToolsList = ((Get-Content $ToolsForInstallFile) -replace ',', '=') -join "`n" | ConvertFrom-StringData
```
I didn't read the spec well enough, so ignore if you already discussed this, but why would the versions.txt file have commas in it? The cmake example above is already in the right format.

natemcmaster commented on eng/common/nativetoolsbootstrap.ps1:
```diff
@@ -0,0 +1,129 @@
+#>
+[CmdletBinding(PositionalBinding=$false)]
+Param (
+    [string] $AzureStorageUrl = "https://dotnetfeed.blob.core.windows.net",
```
Might be good to remove Azure out of the parameter names. Azure happens to be the service we use to host these files, but in theory anyone could setup a remote tools feed, right?

natemcmaster commented on eng/common/native/install-cmake.ps1:
```diff
@@ -0,0 +1,104 @@
+    $ToolNameMoniker = "$ToolName-$Version-$ToolOs-$Arch"
+
+    # Install tool
+    if ((Test-Path $$AssemblyPath) -And (-Not $Force)) {
```
Typo: looks like you meant `$AssemblyPath`. Also, it's not defined yet.

chcosta commented on eng/common/nativetoolsbootstrap.ps1:
```diff
@@ -0,0 +1,129 @@
+    $ToolsList.GetEnumerator() | ForEach-Object {
+        $ToolName = $_.Key
+        $ToolVersion = $_.Value
+        $InstallerFilename = "install-$ToolName.ps1"
```
This script doesn't know the names of the artifacts to download.  The install scripts derive the names.

chcosta replied:
Each tool is going to have its own format for naming its artifacts. ie

```
cmake-3.11.1-win32-x86.zip 
cmake-3.11.1-win64-x64.zip 
```
vs

```
python-3.6.5-embed-amd64.zip 
python-3.7.0b3-embed-win32.zip 
```

It's not something you can auto-derive unless we repackage / rename all of the assets.

weshaggard commented on eng/common/native/install-cmake.ps1:
```diff
@@ -0,0 +1,99 @@
+        $ToolOs = "win32"
+    }
+    $ToolNameMoniker = "$ToolName-$Version-$ToolOs-$Arch"
+    $AssemblyPath = Join-Path $InstallPath "$ToolName\$Version\$ToolNameMoniker\bin\$ToolName.exe"
```
I would call this `ToolPath` or `ToolFilePath` as opposed to assembly path.

weshaggard commented on eng/common/nativetoolsbootstrap.cmd:
```diff
@@ -0,0 +1,3 @@
+@echo off
```
Do you want to intentionally call this "nativetoolsbootstrap"? Why not just "init-tools"?

weshaggard commented on eng/common/nativetoolsbootstrap.ps1:
```diff
@@ -0,0 +1,124 @@
+Entry point script for installing native tools
+
+.DESCRIPTION
+Reads $RepoRoot\eng\NativeToolsVersion.txt file to determine native assets to install
```
I might suggest that "NativeToolsVersion.txt" be a parameter as well, which can be defaulted but can be overridden.

weshaggard commented on eng/common/nativetoolsbootstrap.ps1:
```diff
@@ -0,0 +1,124 @@
+Switch specifying to not install anything, but cleanup native asset folders
+
+.PARAMETER Force
+Force install of tools even if they previously exist
```
Does force overwrite or clean existing tools? I would suggest it clean first.

markwilkie commented:
> > What about using hash compare for version detection since we don't need to determine version order?haggard 

> I don't understand how this would help? This would only work if you want to do exact matches. Generally we are going to want to ask the question do we already have a this version or higher.

We don't need to determine if a version is higher or not, just if it's different.  My understanding is that the bootstrapping scripts will request a specific version, and if that version is not the one being used, then it should be - even if this means going backward.

I'd much rather land in a place where we're using a "match" vs. "no match" then land in the quagmire of version compares and re implement parts of msi....   :(

@weshaggard @chcosta

weshaggard commented:
>We don't need to determine if a version is higher or not, just if it's different. My understanding is that the bootstrapping scripts will request a specific version, and if that version is not the one being used, then it should be - even if this means going backward.

While that might work for official builds it isn't going to fly for dev builds. Devs are going to quickly get annoyed when they have a newer version of something on their machine which will work yet we keep downloading one for their build, and doing so for every repo they have on their machine. (we should think about how we can share these across repos).

chcosta commented:
We can install all builds into the global user profile, but make official builds force install (clean then install).  We can optionally provide an env var override if someone wants to install somewhere else.

weshaggard commented:
>For dev builds we can put native tools into the global user profile

That can help but that implies we can control the global install. How does that work for folks that install cmake via installer or VS via installer?

chcosta commented:
Spoke with @weshaggard .  We agree that this needs to be sorted, and designed before mass rollout, but I'd like to tackle that issue of version detection / use separately.  Issue https://github.com/dotnet/arcade/issues/141 is tracking this work. I'll move some of the comments there and continue the discussion, but not block on that.

tmat commented on eng/common/init-tools-native.cmd:
```diff
@@ -0,0 +1,3 @@
+@echo off
+powershell -NoProfile -NoLogo -ExecutionPolicy ByPass %~dp0init-tools-native.ps1 %*
```
Use 
`powershell -NoProfile -NoLogo -ExecutionPolicy ByPass -command "& """%~dp0init-tools-native.ps1""" %*"` 

this will handle potential spaces in paths.

BTW, do we actually need the .cmd wrapper at all? Why not just have .ps1 file?

weshaggard replied:
>BTW, do we actually need the .cmd wrapper at all? Why not just have .ps1 file?

I do think the wrappers are helpful for at least the entry point scripts as most people work out of a cmd window. For the other scripts that we don't expect folks to call directly I think we could consider just having a ps1 script.

tmat replied:
I agree. I wouldn't think this script needs to be called by people. The only scripts I should be calling is `Restore.cmd`, `Build.cmd`, etc. They would make sure that native tools are initialized by calling this script.

chcosta replied:
Yes, I need to wire this up into the build system so it plays nicely with our other scripts.  Having the wrapper script is helpful (to me), and it's possible that some other ongoing work at the moment (DevTestLab), may make use of such a script.

tmat commented:
Re using `global.json` to list the versions of native tools.

I already do so in RepoToolset for `vswhere`, see https://github.com/dotnet/interactive-window/blob/master/global.json.

weshaggard commented on eng/common/init-tools-native.ps1:
```diff
@@ -0,0 +1,133 @@
+Directory to install native toolset.  This is a command-line override for the default
+Install directory precedence order:
+- InstallDirectory command-line override
+- COMMONLIBRARY_NATIVEINSTALLDIRECTORY environment variable
```
Why call this `COMMONLIBRARY_NATIVEINSTALLDIRECTORY` For consistency with the folder name of the default I would use `NETCOREENG_INSTALLDIRECORY` or something along those lines as I do think we may want to share this across other eng directories, like dotnet-install.

tmat replied:
Also, separate words by `_`: `INSTALLDIRECTORY` -> `INSTALL_DIRECTORY`

weshaggard commented on eng/common/init-tools-native.ps1:
```diff
@@ -0,0 +1,133 @@
+    $InstallBin = Join-Path $NativeBaseDir "bin"
+
+    if ($Clean -Or $Force) {
+        Write-Host "Cleaning '$NativeBaseDir'"
```
Do we really want the clean/force to remove the entire directory even though it might be shared across repos?

weshaggard replied:
We may want to consider scoping it to just the particular tool version.

chcosta replied:
Agreed, I had the same thought after our conversation yesterday.

weshaggard commented on eng/common/init-tools-native.ps1:
```diff
@@ -0,0 +1,133 @@
+    $ToolsList = ((Get-Content $ToolsVersionsFile) -replace ',','=') -join "`n" | ConvertFrom-StringData
+
+    Write-Verbose "Required native tools:"
+    $ToolsList.GetEnumerator() | ForEach-Object {
```
Is there a particular reason you are explicitly calling `GetEnumerator`?

chcosta replied:
No, and... all of this has changed as I'm switching this to parse global.json.

weshaggard commented on eng/common/init-tools-native.ps1:
```diff
@@ -0,0 +1,133 @@
+    }
+    $ToolsList = ((Get-Content $ToolsVersionsFile) -replace ',','=') -join "`n" | ConvertFrom-StringData
+
+    Write-Verbose "Required native tools:"
```
We may need to consider someway of marking tools as optional.

chcosta replied:
Filed https://github.com/dotnet/arcade/issues/149

weshaggard commented on eng/common/init-tools-native.ps1:
```diff
@@ -0,0 +1,133 @@
+
+    # Execute installers
+    Write-Host "Executing installers"
+    $ToolsList.GetEnumerator() | ForEach-Object {
```
If you want to look more code like you can make this `foreach($tool in $ToolList)` and then just use `$tool` instead of `$_` everywhere.

weshaggard commented on eng/common/native/CommonLibrary.psm1:
```diff
@@ -0,0 +1,342 @@
+    # Define verbose switch if undefined
+    $Verbose = $VerbosePreference -Eq "Continue"
+    
+    $TempDir = CommonLibrary\Get-TempDirectory
```
Any particular reason you are using uppercase letters in your variable names for the first letter? I tend to use C# variable style which is pascal casing.

chcosta replied:
It's what the working group agreed on when Arcade started development and was looking at PS conventions.

weshaggard replied:
OK it would be good to try and capture that somewhere in here so we can be consistent and point to it whenever folks are making changes.

weshaggard commented on eng/common/native/CommonLibrary.psm1:
```diff
@@ -0,0 +1,342 @@
+        while($Attempt -Lt $DownloadRetries)
+        {
+            try {
+                Invoke-WebRequest -UseBasicParsing -Uri $Uri -OutFile $Path
```
We are going to end up needing proxy server support. See https://github.com/dotnet/corefx/pull/23467 for example

chcosta replied:
Filed https://github.com/dotnet/arcade/issues/150

tmat commented on eng/common/native/install-cmake.ps1:
```diff
@@ -0,0 +1,119 @@
+# Import common library modules
+Import-Module -Name (Join-Path $CommonLibraryDirectory "CommonLibrary.psm1")
+
+try {
```
nit: build.ps1 uses 2 spaces for indentation, this file uses 4

tmat replied:
Perhaps we should add .editorconfig file with the setting

---
In reply to: [187726050](https://github.com/dotnet/arcade/pull/142#discussion_r187726050) [](ancestors = 187726050)

tmat commented on eng/common/build.ps1:
```diff
@@ -90,6 +90,14 @@ function MakeGlobalSdkAvailableLocal {
+function InstallNativeTools {
+  $NativeToolsInstaller = Join-Path $PSScriptRoot "init-tools-native.ps1"
+
+  if (Test-Path $NativeToolsInstaller) {
```
>if (Test-Path $NativeToolsInstaller) { [](start = 2, length = 38)

The files in common are gonna be always present, no?

tmat replied:
I think this should instead be conditioned on presence of "native-tools" entry in global.json file

---
In reply to: [187726346](https://github.com/dotnet/arcade/pull/142#discussion_r187726346) [](ancestors = 187726346)

tmat commented on eng/common/init-tools-native.ps1:
```diff
@@ -0,0 +1,126 @@
+Install directory precedence order:
+- InstallDirectory command-line override
+- NETCOREENG_INSTALL_DIRECTORY environment variable
+- (default) %USERPROFILE%/.netcoreeng/native
```
Note that we use DOTNET_INSTALL_DIR to find dotnet cli. 
I guess another reason why dotnet CLI should be special and not treated like other native tools.

chcosta replied:
Filed https://github.com/dotnet/arcade/issues/151

weshaggard commented:
Did you intentionally remove the global.json file changes, given that this repo doesn't need cmake? I like the format of including these versions under "native-tools" we may need to abstract that to get a list of groups like "sdk", "native-tools", etc but lets see how that happens when we include dotnet cli in the mix. 

Overall the changes look good.

chcosta commented:
Yes, I removed the changes to "global.json" from this PR since Arcade isn't using it.  I updated the design doc with an example of what global.json would look like https://github.com/dotnet/arcade/pull/131/files#diff-4188151404bfe2cc4f931c8e5fa19b1dR108

I've also changed the Azure container name from `chcosta-test` to `netcoreeng` 

```TEXT
https://dotnetfeed.blob.core.windows.net/netcoreeng/native-assets
```

weshaggard commented on eng/common/init-tools-native.ps1:
```diff
@@ -0,0 +1,126 @@
+  [switch] $Force = $False,
+  [int] $DownloadRetries = 5,
+  [int] $RetryWaitTimeInSeconds = 30,
+  [string] $GlobalJsonFile = "$PSScriptRoot\..\..\global.json"
```
We might want to be careful about relative path structure. Perhaps we should search-up each directory until we find it which is what other tools do.

chcosta replied:
Agree.  I'll look into making this change with my Linux work.

chcosta commented on eng/common/build.ps1:
```diff
@@ -90,6 +90,15 @@ function MakeGlobalSdkAvailableLocal {
+function InstallNativeTools {
+  $NativeToolsInstaller = Join-Path $PSScriptRoot "init-tools-native.ps1"
+
+  $NativeTools = $GlobalJson | Select-Object -Expand "native-tools" -ErrorAction SilentlyContinue
```
This particular statement isn't re-parsing the json file, it's trying to grab the "native-tools" Property from the `GlobalJson` Powershell object that was already parsed.  We could also move this check to the main loop, but given that `GlobalJson` is a global object, I put the check here.

Unless you're talking about passing the parsed Json object into the `init-tools-native.ps1` script.  I suppose that could be an option, but I stayed away from it in case someone used the `init-tools-native.cmd` entry point because then the script would need to parse "global.json" anyways.

tmat replied:
I think this is better. It is ok to reparse. 

BTW, I think you can shorten this to: `$GlobalJson.'native-tools'`

---
In reply to: [187738228](https://github.com/dotnet/arcade/pull/142#discussion_r187738228) [](ancestors = 187738228)

tmat replied:
On second thought I'm not sure if that throws or returns $null on failure. So maybe not.

---
In reply to: [187747123](https://github.com/dotnet/arcade/pull/142#discussion_r187747123) [](ancestors = 187747123,187738228)
